drivers/can/can_mcp25xxfd.c

```diff
@@ -0,0 +1,1792 @@
+	}
+
+	dev_data->mode = mode;
+
+	return 0;
+}
+
+static int mcp25xxfd_set_timing(const struct device *dev, const struct can_timing *timing,
+				bool is_nominal)
+{
+	struct mcp25xxfd_data *dev_data = dev->data;
+	struct mcp25xxfd_nbtcfg *nominal_timing_reg;
+	struct mcp25xxfd_dbtcfg *data_timing_reg;
+	uint8_t *sjw_cached;
+	uint16_t addr_timing_reg, reg_size;
+	int ret;
+
+	if (!timing) {
+		return -EINVAL;
+	}
+
+	/* we have to be in a configuration mode at this point */
+	if (dev_data->started) {
+		return -EBUSY;
+	}
+
+	__ASSERT_NO_MSG(timing->prop_seg == 0);
+	if (is_nominal) {
+		__ASSERT_NO_MSG(timing->sjw == CAN_SJW_NO_CHANGE ||
+				(timing->sjw >= 1 && timing->sjw <= 128));
+		__ASSERT_NO_MSG(timing->phase_seg1 >= 2 && timing->phase_seg1 <= 256);
+		__ASSERT_NO_MSG(timing->phase_seg2 >= 1 && timing->phase_seg2 <= 128);
+		__ASSERT_NO_MSG(timing->prescaler >= 1 && timing->prescaler <= 256);
+	} else {
+		__ASSERT_NO_MSG(timing->sjw == CAN_SJW_NO_CHANGE ||
+				(timing->sjw >= 1 && timing->sjw <= 16));
+		__ASSERT_NO_MSG(timing->phase_seg1 >= 1 && timing->phase_seg1 <= 32);
+		__ASSERT_NO_MSG(timing->phase_seg2 >= 1 && timing->phase_seg2 <= 16);
+		__ASSERT_NO_MSG(timing->prescaler >= 1 && timing->prescaler <= 256);
+	}
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	sjw_cached = is_nominal ? &dev_data->sjw : &dev_data->sjw_data;
+	if (timing->sjw != CAN_SJW_NO_CHANGE) {
+		*sjw_cached = timing->sjw;
+	}
+
+	addr_timing_reg = is_nominal ?  MCP25XXFD_REG_NBTCFG : MCP25XXFD_REG_DBTCFG;
+
+	if (is_nominal) {
+		nominal_timing_reg = mcp25xxfd_get_spi_buf_ptr(dev, sizeof(*nominal_timing_reg));
+		nominal_timing_reg->BRP = timing->prescaler - 1;
+		nominal_timing_reg->TSEG1 = timing->prop_seg + timing->phase_seg1 - 1;
+		nominal_timing_reg->TSEG2 = timing->phase_seg2 - 1;
+		nominal_timing_reg->SJW = *sjw_cached - 1;
+		reg_size = sizeof(*nominal_timing_reg);
+	} else {
+		data_timing_reg = mcp25xxfd_get_spi_buf_ptr(dev, sizeof(*data_timing_reg));
+		data_timing_reg->BRP = timing->prescaler - 1;
+		data_timing_reg->TSEG1 = timing->prop_seg + timing->phase_seg1 - 1;
+		data_timing_reg->TSEG2 = timing->phase_seg2 - 1;
+		data_timing_reg->SJW = *sjw_cached - 1;
+		reg_size = sizeof(*data_timing_reg);
+	}
+
+	ret = mcp25xxfd_write_reg(dev, addr_timing_reg, reg_size);
+	if (ret < 0) {
+		LOG_ERR("Failed to write device configuration [%d]", ret);
+		goto done;
+	}
+
+#ifdef CONFIG_CAN_FD_MODE
+	if (!is_nominal) {
+		dev_data->tdco = timing->prescaler * (timing->prop_seg + timing->phase_seg1);
+	}
+#endif
+
+done:
+	k_mutex_unlock(&dev_data->mutex);
+
+	return ret;
+}
+
+static int mcp25xxfd_set_timing_nominal(const struct device *dev, const struct can_timing *timing)
+{
+	return mcp25xxfd_set_timing(dev, timing, true);
+}
+
+
+#if defined(CONFIG_CAN_FD_MODE)
+static int mcp25xxfd_set_timing_data(const struct device *dev, const struct can_timing *timing)
+{
+	return mcp25xxfd_set_timing(dev, timing, false);
+}
+#endif
```

Comment:
Why not just get rid of `is_nominal` in `mcp25xxfd_set_timing()` and move the relevant code to each of these functions? That would make the code more readable.

Reply:
I thought this way would remove code duplication. But I'll look over the part.